### PR TITLE
Allow mixing dask and numpy arrays in @guvectorize

### LINKED ISSUE #6855: Mixing of dask.array and numpy.ndarray in Numba @guvectorize
It appears that `dask.array` and `numpy.ndarray` cannot be mixed when using Numba's `@guvectorize` decorator - see the minimal example below. It would be a nice feature if dask could resolve this (or at least raise a more helpful exception!). Regardless, this is a truly helpful package!

**Minimal Example**
```python
import dask.array as da
import numpy as np
import numba as nb

# large array
x = np.random.random((300000, 100, 10))
y = da.from_array(x) # for the sake of the example

# small vector
v = np.ones(10)
w = da.from_array(v)

@nb.guvectorize([nb.void(nb.f8[:, :], nb.f8[:], nb.f8[:])],
                '(m, n), (n) -> (m)', target='parallel')
def test(arr, vec, out):
    # really just a test function
    out[:] = (arr + vec).sum(axis=1)

out1 = test(x, v) # works
out2 = test(y, w).compute() # works

out3 = test(y, v) # fails with ValueError
```
The following `ValueError` is raised ... (on dask version 2.30.0)
```
[...]
ValueError: `dtype` inference failed in `apply_gufunc`.

Please specify the dtype explicitly using the `output_dtypes` kwarg.

Original error is below:
------------------------
ValueError('test: Input operand 1 has a mismatch in its core dimension 0, with gufunc signature (m, n), (n) -> (m) (size 10 is different from 1)',)

Traceback:
---------
  File "<...>/miniconda3-latest/envs/<...>/lib/python3.6/site-packages/dask/array/core.py", line 377, in apply_infer_dtype
    o = func(*args, **kwargs)
```

### PR DESCRIPTION
**Pull request by jsignell**
- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask`
- [x] Closes #6855 

I am not sure that this is the right time for adding a meta_nonempty for dask.array but I thought it might be? Also I suspect that this'll have repercussions for cupy @quasiben.

### CODE REVIEW

**Comment by martindurant** [comment]
ping @quasiben or maybe suggest someone else to review

File: dask/array/utils.py [hunk 1]
```diff
@@ -28,6 +28,22 @@ def normalize_to_array(x):
...
+    ----------
+    x: array-like
+        An object that looks sufficiently like a Numpy array
+    Returns
+    -------
+    array-like with one element of the correct dtype and ndims
+    """
+    if not is_arraylike(x):
+        return x
+    return np.ones((1,) * x.ndim, dtype=x.dtype)
```
Comment (quasiben) [inline dask/array/utils.py:44]: Perhaps this should use [np.ones_like[(https://numpy.org/doc/stable/reference/generated/numpy.ones_like.html) instead.  cc @pentschev

File: dask/array/core.py [hunk 2]
```diff
@@ -371,10 +371,7 @@ def apply_infer_dtype(func, args, kwargs, funcname, suggest_dtype="dtype", nout=
     : dtype or List of dtype
         One or many dtypes (depending on ``nout``)
     """
-    args = [
-        np.ones((1,) * x.ndim, dtype=x.dtype) if isinstance(x, Array) else x
-        for x in args
-    ]
+    args = [meta_nonempty(x) for x in args]
```
Comment (pentschev) [inline dask/array/core.py:374]: Wouldn't using `meta_from_array` suffice here?

File: dask/array/utils.py (hunk 1 above)
Comment (pentschev) [reply dask/array/utils.py:44]: If this is indeed necessary, we would probably want https://github.com/dask/dask/blob/8f9fc1635170cda07a4bb8c4bd93e96da151d938/dask/array/utils.py#L340-L349 here, but I'm wondering why can't we just use `meta_from_array` instead.

File: dask/array/core.py (hunk 2 above)
Comment (jsignell) [reply dask/array/core.py:374]: My understanding is it needs to be a nonempty array for gufuncs to work. If I switch this to `meta_from_array` I get:

```
           ValueError: `dtype` inference failed in `apply_gufunc`.
           
           Please specify the dtype explicitly using the `output_dtypes` kwarg.
           
           Original error is below:
           ------------------------
           ValueError('cannot call `vectorize` on size 0 inputs unless `otypes` is set')
           
           Traceback:
           ---------
             File "/home/julia/dask/dask/array/core.py", line 377, in apply_infer_dtype
               o = func(*args, **kwargs)
             File "/home/julia/conda/envs/dask-dev/lib/python3.8/site-packages/numpy/lib/function_base.py", line 2091, in __call__
               return self._vectorize_call(func=func, args=vargs)
             File "/home/julia/conda/envs/dask-dev/lib/python3.8/site-packages/numpy/lib/function_base.py", line 2157, in _vectorize_call
               res = self._vectorize_call_with_signature(func, args)
             File "/home/julia/conda/envs/dask-dev/lib/python3.8/site-packages/numpy/lib/function_base.py", line 2226, in _vectorize_call_with_signature
               raise ValueError('cannot call `vectorize` on size 0 inputs '

```

Comment (pentschev) [reply dask/array/core.py:374]: In that case, couldn't you just replace `np.ones` by [`ones_like_safe`] in the old code? That way you wouldn't even need the new `meta_from_array` function, unless I'm missing something.

**Comment by jsignell** [comment]
Thanks for the feedback @pentschev. I think I had made it overly complicated for no reason.

**Review by pentschev — APPROVED** [review]
Thanks @jsignell , this reads much better, and looks good to me now! 😄

**Comment by jsignell** [comment]
I am going to merge this if it passes.

File: dask/array/utils.py [hunk 3]
```diff
@@ -416,8 +416,10 @@ def ones_like_safe(a, shape, **kwargs):
     is supported (requires NumPy >= 1.17), otherwise falls back to
     using the old behavior, returning ``np.ones(shape, **kwargs)``.
     """
+    from .creation import ones_like
+
     try:
-        return np.ones_like(a, shape=shape, **kwargs)
+        return _array_like_safe(np.ones_like, ones_like, a, shape=shape, **kwargs)
```
Comment (jsignell) [inline dask/array/utils.py:422]: I think this is a neccessary improvement on this function, but @pentschev please correct me if this is wrong.

Comment (pentschev) [reply dask/array/utils.py:422]: No, this isn't necessary. `ones_like` already existed before we added `like=` to creation functions like `np.asarray`, in fact, the `like=` argument name comes from the `*_like` functions. We have added `like=` to all creation functions, including `np.ones`, but using `np.ones_like` is equivalent to `np.ones(..., like=...)`. Therefore, `ones_like_safe` already works without the change proposed here.

Comment (jsignell) [reply dask/array/utils.py:422]: Ok thank you. I will think more about what I am doing wrong then. I was getting a segfault without this change which made me think that maybe I was hitting a wierd loop or somthing. I

**Comment by jsignell** [comment]
Ok I made `ones_like_safe` operate on the `_meta` if available so that the result is always an evaluated array

**Comment by jsignell** [comment]
@jrbourbeau I plan to merge this tomorrow unless you have any comments.

File: dask/array/core.py [hunk 4]
```diff
@@ -369,8 +369,15 @@ def apply_infer_dtype(func, args, kwargs, funcname, suggest_dtype="dtype", nout=
     : dtype or List of dtype
         One or many dtypes (depending on ``nout``)
     """
+    from .utils import ones_like_safe
+
+    # make sure that every arg is an evaluated array
     args = [
-        np.ones((1,) * x.ndim, dtype=x.dtype) if isinstance(x, Array) else x
+        ones_like_safe(
+            x._meta if isinstance(x, Array) else x, shape=((1,) * x.ndim), dtype=x.dtype
```
Comment (pentschev) [inline dask/array/core.py:377]:
```suggestion
            meta_from_array(x), shape=((1,) * x.ndim), dtype=x.dtype
```

File: dask/array/core.py [hunk 5]
```diff
@@ -369,8 +369,15 @@ def apply_infer_dtype(func, args, kwargs, funcname, suggest_dtype="dtype", nout=
     : dtype or List of dtype
         One or many dtypes (depending on ``nout``)
     """
+    from .utils import ones_like_safe
```
Comment (pentschev) [inline dask/array/core.py:372]:
```suggestion
    from .utils import meta_from_array, ones_like_safe
```

**Review by pentschev — COMMENTED** [review]
Sorry @jsignell for the late reply, overall changes look good, I suggested only using `meta_from_array` to get the correct `_meta`, hopefully that's a simpler way and will work as well.

File: dask/array/core.py (hunk 4 above)
Comment (jsignell) [reply dask/array/core.py:377]: Ah true! I forgot about meta_from_array :facepalm:

**Review by pentschev — APPROVED** [review]
LGTM, thanks @jsignell !

**Comment by jsignell** [comment]
Thanks for all the reviews! :sweat_smile:

**Pull request merged**